x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,294 @@
+        }
+        if (matchCriteria == null || matchCriteria.isEmpty()) {
+            throw new IllegalArgumentException("matchCriteria must not be null or empty");
+        }
+        for (String matchCriteriaKey : matchCriteria.keySet()) {
+            if (ALLOWED_MATCH_CRITERIA.contains(matchCriteriaKey) == false) {
+                throw new IllegalArgumentException("matchCriteria key [" + matchCriteriaKey + "] is not allowed");
+            }
+        }
+        if (Strings.isNullOrEmpty(rulesetId)) {
+            throw new IllegalArgumentException("rulesetId must not be null or empty");
+        }
+
+        this.organicQuery = organicQuery;
+        this.matchCriteria = matchCriteria;
+        this.rulesetId = rulesetId;
+        this.pinnedIds = pinnedIds;
+        this.pinnedIdsSupplier = pinnedIdsSupplier;
+        this.pinnedDocs = pinnedDocs;
+        this.pinnedDocsSupplier = pinnedDocsSupplier;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (pinnedIdsSupplier != null) {
+            throw new IllegalStateException("pinnedIdsSupplier must be null, can't serialize suppliers, missing a rewriteAndFetch?");
+        }
+        if (pinnedDocsSupplier != null) {
+            throw new IllegalStateException("pinnedDocsSupplier must be null, can't serialize suppliers, missing a rewriteAndFetch?");
+        }
+
+        out.writeNamedWriteable(organicQuery);
+        out.writeGenericMap(matchCriteria);
+        out.writeString(rulesetId);
+        if (pinnedIds != null) {
+            out.writeBoolean(true);
+            out.writeStringCollection(pinnedIds);
+        } else {
+            out.writeBoolean(false);
+        }
+        if (pinnedDocs != null) {
+            out.writeBoolean(true);
+            out.writeList(pinnedDocs);
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    public String rulesetId() {
+        return rulesetId;
+    }
+
+    public Map<String, Object> matchCriteria() {
+        return matchCriteria;
+    }
+
+    public QueryBuilder organicQuery() {
+        return organicQuery;
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(ORGANIC_QUERY_FIELD.getPreferredName(), organicQuery);
+        builder.startObject(MATCH_CRITERIA_FIELD.getPreferredName());
+        builder.mapContents(matchCriteria);
+        builder.endObject();
+        builder.field(RULESET_ID_FIELD.getPreferredName(), rulesetId);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
+            throw new IllegalArgumentException("Can't have both pinned ids and pinned docs");
```

Comment:
when is this reached? when multiple pinned query rules match and they have a combination of `ids` and `docs`?

Reply:
Yes, that was what I was trying to guard against. I can try to come up with a demo example of this.

Reply:
Per discussion - I'll leave this error in here for now, but maybe update the error message. This could potentially go away in the future if we make changes to the pinned query builder.